src/services/bunker.py

```diff
@@ -1,15 +1,413 @@
+            # If no one report was made - we should use initial epoch as last completed epoch
+            # Because in this case 'cl_rebase' will contain all rewards from initial epoch
+            last_completed_epoch = self.f_conf.initial_epoch
+            total_last_effective_balance = total_ref_effective_balance
+            total_last_lido_effective_balance = total_ref_lido_effective_balance
+        else:
+            last_completed_epoch = last_member_report_ref_epoch
+            last_all_validators = self.w3.cc.get_validators(last_report_blockstamp.state_root)
+            last_lido_validators = self.w3.lido_validators.filter_lido_validators(self.lido_keys, last_all_validators)
+            total_last_effective_balance = self._calculate_total_effective_balance(last_all_validators)
+            total_last_lido_effective_balance = self._calculate_total_effective_balance(
+                [v.validator for v in last_lido_validators]
+            )
+
+        mean_total_effective_balance = (total_ref_effective_balance + total_last_effective_balance) // 2
+        mean_total_lido_effective_balance = (
+            (total_ref_lido_effective_balance + total_last_lido_effective_balance) // 2
+        )
+
+        epochs_passed = current_ref_epoch - last_completed_epoch
+
+        normal_cl_rebase = self._calculate_normal_cl_rebase(
+            epochs_passed, mean_total_lido_effective_balance, mean_total_effective_balance
+        )
+
+        logger.info({"msg": f"Normal CL rebase: {normal_cl_rebase} Gwei"})
+        return Gwei(normal_cl_rebase)
+
+    def _calculate_nearest_and_far_cl_rebase(self, blockstamp: BlockStamp) -> tuple[Gwei, Gwei]:
+        """
+        Calculate CL rebase from nearest and far epochs to ref epoch given the changes in withdrawal vault
+        """
+        logger.info({"msg": "Calculating nearest and far CL rebase"})
+        nearest_slot = (blockstamp.slot_number - NEAREST_EPOCH_DISTANCE * self.c_conf.slots_per_epoch)
+        far_slot = (blockstamp.slot_number - FAR_EPOCH_DISTANCE * self.c_conf.slots_per_epoch)
+
+        lookup_max_deep = self.c_conf.slots_per_epoch * self.f_conf.epochs_per_frame
+        nearest_blockstamp = get_first_non_missed_slot(self.w3.cc, SlotNumber(nearest_slot), lookup_max_deep)
+        far_blockstamp = get_first_non_missed_slot(self.w3.cc, SlotNumber(far_slot), lookup_max_deep)
+
+        nearest_cl_rebase = self._calculate_cl_rebase_between(nearest_blockstamp, blockstamp)
+        far_cl_rebase = self._calculate_cl_rebase_between(far_blockstamp, blockstamp)
+
+        logger.info({"msg": f"CL rebase {nearest_cl_rebase,far_cl_rebase=}"})
+        return nearest_cl_rebase, far_cl_rebase
+
+    def _calculate_cl_rebase_between(self, prev_blockstamp: BlockStamp, curr_ref_blockstamp: BlockStamp) -> Gwei:
+        """
+        Calculate CL rebase from prev_blockstamp to ref_blockstamp given the changes in withdrawal vault
```

Comment:
Maybe add a bit more details?

Skimmed validator rewards are sent to 0x01 withdrawal credentials address, which in the case of Lido is a withdrawal vault. To account for all changes in validators' balances, we must take its balance into account

Reply:
done